cmds/core/pci/pci_test.go

```diff
@@ -216,17 +208,68 @@ func TestRegisters(t *testing.T) {
 					Config: []byte{0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77},
 				},
 			},
-			cmds:    []string{"0.w=10"},
-			wantErr: "open config: no such file or directory",
+			cmds: []string{"0.w=10"},
+			err:  os.ErrNotExist,
 		},
 	} {
 		t.Run(tt.name, func(t *testing.T) {
 			buf := &bytes.Buffer{}
 			log.SetOutput(buf)
-			registers(tt.devices, tt.cmds...)
-			if !strings.Contains(buf.String(), tt.wantErr) {
-				t.Errorf("registers() = %q, should contain: %q", buf.String(), tt.wantErr)
+			if err := registers(tt.devices, tt.cmds...); !errors.Is(err, tt.err) {
+				t.Errorf("registers(): got %v, want %v", err, tt.err)
 			}
 		})
 	}
 }
+
+// This test is here because of very strange encoding/json behavior.
+// Right now it works.
+// It stopped working with the JSON file that has been in use for 8 years.
+// How to see it yourself.
+// It seems the json package makes an effort to parse "Primary": "00" to a Uint8, but some usages
+// of the package cause a failure? it's very mysterious.
+//
+// go test, see it pass.
+// go test -test.run TestJSON
+// PASS
+// ok  	github.com/u-root/u-root/cmds/core/pci	0.002s
+// in the JSON file, change the JSON for Primary:
+// +               "Primary": "00",
+// -               "Primary": 55,
+// Watch it fail, even though it has worked for years.
+// --- FAIL: TestJSONWeirdness (0.00s)
+//
+//	pci_test.go:240: json: cannot unmarshal string into Go struct field PCI.Primary of type uint8
+//	pci_test.go:243: json: cannot unmarshal string into Go struct field PCI.Primary of type uint8
+//	pci_test.go:247: run() got json: cannot unmarshal string into Go struct field PCI.Primary of type uint8:JSON parsing failed, want nil
+//
+// Change the true to false on the if. So the code is removed at link time.
+// rminnich@pop-os:~/go/src/github.com/u-root/u-root/cmds/core/pci$ go test -test.run TestJSONWeirdness
+// PASS
+// ok  	github.com/u-root/u-root/cmds/core/pci	0.002s
+// NOTE, the JSON did not change between the bad and good run; only the way in which you called
+// Unmarshal.
+// So, just enabling those lines cause the PREVIOUS Unmarshal's to fail! Precrime!
+// tinygo and go see the same issue.
+// It's very hard to get this to repro, much less make a standalone repro, but hopefully this
+// test will help protect us in future.
+// And, the weirdest part:
+func TestJSONWeirdness(t *testing.T) {
+	var d pci.Devices
+	b, err := os.ReadFile("testdata/testfile1.json")
+	if err != nil {
+		t.Log(err)
+	}
+	if err := json.Unmarshal(b, &d); err != nil {
+		t.Log(err)
+	}
+	if err := json.Unmarshal(b, &d); err != nil {
+		t.Log(err)
+	}
+	if true {
```

Comment:
This really look weird, but why `if true`?

Reply:
it is a way to mark how to exercise the bug I'm seeing json, please see the comment. I'd like to leave it there for that reason, until I can find out what's going on. I'm still working on a repro